Additional Pod Controller Scans

endzyme opened the pull request:
- Refactored the way controllers work to be an interface
- Added configurable controllers to include in scans
- Added daemonsets, jobs and cronjobs in scans

Explicitly did not add ReplicationControllers and ReplicaSets since they are either deprecated or controlled by deployments (generally).

- Added tests to supportedController scanner
- Added documentation on the exit codes 
- Added code documentation for longer functions

bobby-brennan commented on README.md:
```diff
@@ -116,6 +116,20 @@ rather than a cluster:
+  <dt>Exit 0</dt>
+  <dd>Successful exit code</dd>
+  <dt>Exit 1</dt>
+  <dd>Could not run audit, or application had a failure while running.</dd>
```
This might be better described as "unexpected error". We seem to use it like an HTTP `500`

bobby-brennan commented on README.md:
````diff
@@ -116,6 +116,20 @@ rather than a cluster:
 polaris --audit --audit-path ./deploy/
 ```
 
+##### Exit Codes for Audits
````
The README is getting a little dense. What do you think about breaking this and the CLI flags into a separate file?

bobby-brennan commented on pkg/config/supportedcontrollers.go:
```diff
@@ -0,0 +1,105 @@
+}
+
+// MarshalJSON manages writing the enum into json data or error on unsupported value
+func (s SupportedController) MarshalJSON() ([]byte, error) {
```
The Marshal and Unmarshal functions don't seem to be used...

bobby-brennan commented on pkg/config/supportedcontrollers.go:
```diff
@@ -0,0 +1,105 @@
+	if stringLookupForSupportedControllers[lowerStr] == Unsupported {
+		return 0, fmt.Errorf("Value ('%v') in configuration was not found in Supported Controllers: (%v)", str, strings.Join(ControllerStrings, ","))
+	}
+	return stringLookupForSupportedControllers[lowerStr], nil
```
Do we need to check if `lowerStr` is in the map?

endzyme replied:
Hmm yes it would appear so. I'll add some tests around this and see what the default behavior is for nil retirns.

endzyme replied:
So this is a weird language behavior - i find it strange but it's supposedly supported :man_shrugging: 

https://blog.golang.org/go-maps-in-action

TL;DR; 
if you have a `map[string]int` and you call a key that doesn't exist - you get the `zero value` default of the _value_ type. So in our case `map[string]SupportedController` -- `SupportedController is an enum which has a value of `int`, and in this case `int`s "`zero value`" is 0. (for strings it's `""` and for `interfaces`, `func`, `struct`, you can get `nil` as the zero value)

bobby-brennan commented on pkg/webhook/validator.go:
```diff
@@ -92,16 +93,27 @@ func (v *Validator) Handle(ctx context.Context, req types.Request) types.Respons
+		if yes := v.Config.CheckIfKindIsConfiguredForValidation(req.AdmissionRequest.Kind.Kind); !yes {
+			logrus.Warnf("Skipping, kind (%s) isn't something we are configured to scan", req.AdmissionRequest.Kind.Kind)
+			// FIXME: Should we be returning an OK response as skipped?
+			return admission.ErrorResponse(http.StatusBadRequest, err)
```
Yes, I think for the webhook this should be an `OK` response for unsupported types. Otherwise (IIUC) if a new controller type is introduced, we'll reject it until Polaris gets updated to support it.

bobby-brennan replied:
Still seeing the `FIXME` here, and looks like it's still `StatusBadRequest`.

endzyme replied:
I think maybe it was a bad browser cache on your side? I am not seeing it

bobby-brennan commented:
You'll need to update the RBAC definitions in the chart to regenerate everything in the `./deploy` folder as well.

endzyme commented:
> You'll need to update the RBAC definitions in the chart to regenerate everything in the `./deploy` folder as well.

For sure, I saw that on your PR for the limit ranges. This is a WIP and I wanted to validate this isn't too over the top or out of line with the requested functionality.

bobby-brennan commented:
FYI, in https://github.com/FairwindsOps/polaris/issues/107 someone mentions that `ReplicationControllers` are still widely in use - maybe it makes sense to add them?

endzyme commented:
@bobby-brennan: I've incorporated the following (per request)

- Added `replicationController` type
- Added more testing (better coverage)
- Fixed webhook integration and support for skipping non-scanned controllers and other kinds

bobby-brennan commented on README.md:
```diff
@@ -22,16 +22,19 @@ problems in the future. Polaris can be run in a few different modes:
 - A command-line audit that can be incorporated into your CI/CD pipeline
 
-**Want to learn more?** ReactiveOps holds [office hours on Zoom](https://zoom.us/j/242508205) the first Friday of every month, at 12pm Eastern. You can also reach out via email at `opensource@fairwinds.com`
+**Want to learn more?** FairWinds holds [office hours on Zoom](https://zoom.us/j/242508205) the first Friday of every month, at 12pm Eastern. You can also reach out via email at `opensource@fairwinds.com`
```
`s/W/w/`

bobby-brennan commented on pkg/kube/resources.go:
```diff
@@ -173,6 +199,18 @@ func addResourceFromString(contents string, resources *ResourceProvider) error {
 		err = decoder.Decode(&dep)
 		resources.StatefulSets = append(resources.StatefulSets, dep)
+	} else if resource.Kind == "DaemonSet" {
+		dep := appsv1.DaemonSet{}
```
nit: I clearly copy/pasta'd the code for StatefulSet - `dep` was originally meant to refer to `Deployment`. Maybe change `dep` to `ctrl` everywhere?

endzyme replied:
Ah sounds good. Yeah I guess dep didn't connect for me :) I'll fix that up

bobby-brennan commented:
Looking great Nick! A few comments but I think this is just about ready.

I really like using the interface for different controller types, makes a lot more sense than what I had.

As a side note, we should start testing the webhook for correctness and feature parity. Opened https://github.com/FairwindsOps/polaris/issues/178 to track.

bobby-brennan commented on README.md:
```diff
@@ -22,16 +22,19 @@ problems in the future. Polaris can be run in a few different modes:
+**Want to learn more?** Fairwinds holds [office hours on Zoom](https://zoom.us/j/242508205) the first Friday of every month, at 12pm Eastern. You can also reach out via email at `opensource@fairwinds.com`
 
-## Quickstart
+# Quickstart Dashboard
```
Can we rename this to `Dashboard Quickstart`?

bobby-brennan commented on pkg/dashboard/helpers.go:
```diff
@@ -16,14 +16,14 @@ package dashboard
 )
 
 func getAllControllerResults(nr validator.NamespaceResult) []validator.ControllerResult {
 	results := []validator.ControllerResult{}
```
You can drop this line and add change to `:=` below
